platform/engineering/design-docs/2020-04-16-integration-testing-with-consumer-driven-contract-tests.md

```diff
@@ -0,0 +1,304 @@
+           'Content-Type': 'application/json; charset=utf-8',
+         },
+         body: {
+           // The minimum expected response body.
+           data: { ... },
+           meta: { ... },
+         },
+       },
+     };
+     ```
+     - **The `state` attribute will be used as a key for the API to set up the corresponding state when verifying the contract.** For example, this might entail spinning up a mock third-party service, seeding the database, or simulating a logged in session.
+     - **The body does not need to be comprehensive.** At a minimum, the expected response only needs to define the attributes that are relevant to the test. This means that any extra, unused, or unexpected properties of the response will not break the verification later.
+   - Invoke the relevant methods that make the request being tested.
+   - Make assertions on the expected response.
+
+   **To reiterate, the tests should focus on the components or functions that make requests to the API.** It's not meaningful to simply test a fetch call to an endpoint.
+
+   There is some boilerplate in declaring the mock server and some `before`/`after`/`afterEach` hooks that perform some Pact-specific ceremony. We could simplify all of that boilerplate by writing a reusable helper function (`contractTest()`) that can be invoked like this:
+   ```
+   contractTest('VA.gov Search', 'VA.gov API', mockApi => {
+     describe('GET /search', () => {
+       context('with no results', () => {
+         it('responds with success', async () => {
+           await mockApi.addInteraction(INTERACTIONS.searchWithNoResults);
+           ...
+         })
+       })
+     })
+   });
+   ```
+2. The contract tests will run in continuous integration (CI) when FE changes are pushed.
+   - **If they pass, they generate pacts in the form of JSON files in the `pacts` directory.** JSON files are generated for each pairing of consumer (the FE application) and provider (`vets-api`).
+   - If they fail, it's possible the request wasn't picked up by the mock server because request defined in the interaction didn't match the request that was actually made. It's also possible the component or function that made the request failed an assertion.
+
+   The assumption henceforth is that the CI workflow is done in Circle.
+3. **The CI job invokes a script to publish the pacts to a broker.** The command to publish should require a version to label the consumer and optionally accept tags. The published pacts should be versioned to the current commit hash of `vets-website`. For builds to be deployed, the pacts should also be tagged with the appropriate environment. *Tagging is important to ensure that we verify only the pact associated with the build that is getting deployed and not pacts generated from other branches.*
+
+   This assumes that the broker is accessible from CI. There may be some work involved in configuring the the broker to allow access depending on where it's hosted.
```

Comment:
this seems pretty handwavey for a design doc. can you propose answers to the questions of:
* whether we'll self-host a broker or not, and (if so) where/how in our infrastructure it will live (may require discussion with Ops)
* if we're confident that the broker would be accessible from Circle